### DISCUSSION TITLE
Bash completion scripts directory

### DISCUSSION THREAD

**Discussion by lelegard** [Homebrew/discussions#2575] [Everyday usage] [answered]
Hi,

Could someone explain what is the right place to use for bash completion scripts in Homebrew?

By "bash completion script", I mean a script which defines the appropriate completions for one or more shell commands using the shell builtin `complete`. When a product is installed, it will typically install its completion script here.

In old installations of Linux, all completion scripts are stored in `/etc/bash_completion.d`. They are all executed when the shell starts. All these scripts delay the shell startup and pollute the shell environment with zillions of completion functions you will probably never use.

In recent Linux distros, the completion scripts are stored in `/usr/share/bash-completion/completions`. There is exactly one script (or symbolic link to a script) per shell command. Nothing is loaded on shell startup. When TAB is hit on one command and no completion is defined yet for that command, bash looks for a script (or link) with the same name as the command in `/usr/share/bash-completion/completions`. When the script exists, it is executed. As a consequence, completions are defined on demand, when used for the first time only for each command.

In the latter case, `/etc/bash_completion.d` still exists but is used only for local customization and a handful of legacy products.

I try to figure out how it works on Homebrew and it does not seem obvious:
- Most completion scripts are in `/usr/local/etc/bash_completion.d`. On my Mac, there are 194 scripts for 201 installed formulae. They are executed on shell startup, from `/usr/local/etc/profile.d/bash_completion.sh` if you want bash completion.
- In `/usr/local/share/bash-completion/completions`, I find 8 scripts (`cmake cpack ctest gapplication gdbus gio gresource gsettings`) which seem to be unused. Their definitions are ineffective.

So, even with a quite recent shell (bash 5.1.12), it seems that the old method is still used. Can anyone confirm? How could Homebrew's bash could switch to the new "on-demand" method?

Why are there still 8 scripts in the new directory? Are they installation errors in their respective product? Maybe they assume that the new "on-demand" method is used everywhere and their installation should be fixed on Homebrew.

Thanks for any hint.

**Comment by Bo98** [comment, marked as answer]
It's worth noting that the default Bash on macOS is Bash 3.2 and that is what we primarily support.

Bash 3.2 requires bash-completion v1. Bash 5 can use bash-completion v2 (supplied by the `bash-completion@2` formula). The on demand loading system is a feature of bash-completion v2.

> Why are there still 8 scripts in the new directory? Are they installation errors in their respective product?

It's possible some are installation errors. But it's also possible some of them only work with `bash-completion@2`. They would need to be investigated individually.

**Reply by carlocab** [reply]
On my system:
```
Permissions Size User     Date Modified Name
lrwxr-xr-x    68 carlocab 19 Nov 17:08  cmake -> ../../../Cellar/cmake/3.22.0/share/bash-completion/completions/cmake
lrwxr-xr-x    68 carlocab 19 Nov 17:08  cpack -> ../../../Cellar/cmake/3.22.0/share/bash-completion/completions/cpack
lrwxr-xr-x    68 carlocab 19 Nov 17:08  ctest -> ../../../Cellar/cmake/3.22.0/share/bash-completion/completions/ctest
lrwxr-xr-x    74 carlocab  2 Nov 21:41  gapplication -> ../../../Cellar/glib/2.70.1/share/bash-completion/completions/gapplication
lrwxr-xr-x    67 carlocab  2 Nov 21:41  gdbus -> ../../../Cellar/glib/2.70.1/share/bash-completion/completions/gdbus
lrwxr-xr-x    65 carlocab  2 Nov 21:41  gio -> ../../../Cellar/glib/2.70.1/share/bash-completion/completions/gio
lrwxr-xr-x    71 carlocab  2 Nov 21:41  gresource -> ../../../Cellar/glib/2.70.1/share/bash-completion/completions/gresource
lrwxr-xr-x    71 carlocab  2 Nov 21:41  gsettings -> ../../../Cellar/glib/2.70.1/share/bash-completion/completions/gsettings
lrwxr-xr-x    78 carlocab  4 Mar 19:26  openconnect -> ../../../Cellar/openconnect/8.10/share/bash-completion/completions/openconnect
```
These look like they were all installed by whatever the relevant formula's equivalent of `make install` is.

**Comment by lelegard** [comment]
> Bash 3.2 requires bash-completion v1. Bash 5 can use bash-completion v2 (supplied by the bash-completion@2 formula). The on demand loading system is a feature of bash-completion v2.

OK, so `bash-completion` is for the macOS-provided bash 3 while `bash-completion@2` is for the Homebrew-provided bash 5.

If a formula wants to provide on-demand completion scripts in `/usr/local/share/bash-completion/completions`, it should say `depends_on "bash-completion@2"`, right? Since this forces a dependency on `bash`, it forces the installation of bash 5 from Homebrew. Of course, the completion would work only on this bash, not on the default bash 3.2.

Does it sound reasonable or too greedy?

Shouldn't this be added to the formulae which already install these 8 (for me) or 9 (for you) scripts in `/usr/local/share/bash-completion/completions`?

**Reply by carlocab** [reply]
These formulae will still work even if you don't have `bash` or `bash-completion@2` installed, so auto-installing `bash` seems a bit much. It's likely that most users won't even use the bundled bash completion -- they'll probably be on zsh since that's now the default shell.
